### ISSUE TITLE
Battery Plugin not working Archlinux 4.6.4 (64bits)

### ISSUE THREAD

**Issue by YoEight** [jaor/xmobar#271]
Greetings,

Battery plugin keeps displaying `N/A`. 

My definition:
```
Run Battery       [ "--template" , "Batt: <acstatus>"
                             , "--Low"      , "10"        -- units: %
                             , "--High"     , "80"        -- units: %
                             , "--low"      , "red"
                             , "--normal"   , "orange"
                             , "--high"     , "green"

                             , "--" -- battery specific options
                                       -- discharging status
                                       , "-o"	, "<left>% (<timeleft>)"
                                       -- AC "on" status
                                       , "-O"	, "<fc=orange>Charging</fc>" -- dAA520
                                       -- charged status
                                       , "-i"	, "<fc=green>Charged</fc>" -- #006000
                             ] 50
```

```
➜  ~ xmobar --version
Fields missing from config defaulted: additionalFonts,bgColor,fgColor,sepChar,border,borderColor,textOffset,iconOffset,overrideRedirect,pickBroadest,lowerOnStart,iconRoot,alpha
xmobar 0.24.1
 (C) 2007 - 2010 Andrea Rossato 
 (C) 2010 - 2014 Jose A Ortega Ruiz
 <xmobar@projects.haskell.org>

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the License for more details.
```

```
➜  ~ xmonad --version
xmonad 0.12
```

I really don't have any clue so sorry for not giving more information. 

Regards.

**Comment by dsemi** [comment]
I have the same issue, <acstatus> always appears as "N/A".  Also on Arch Linux.

Broken in xmobar-0.24-1, working in xmobad-0.23.1-19.

**Comment by jaor** [comment]
what's the contents of your directory `/sys/class/power_supply/BAT0`?

**Comment by dsemi** [comment]
alarm
capacity
capacity_level
cycle_count
device -> ../../../PNP0C0A:00
energy_full
energy_full_design
energy_now
manufacturer
model_name
power
power_now
present
serial_number
status
subsystem -> ../../../../../../../../../class/power_supply
technology
type
uevent
voltage_min_design
voltage_now

**Comment by jaor** [comment]
hmm, nothing seems missing or wrong.  could you please show me the
output of

       for f in /sys/class/power_supply/BAT0/*; do echo $f; cat $f; done

?

**Comment by dsemi** [comment]
/sys/class/power_supply/BAT0/alarm
1016000
/sys/class/power_supply/BAT0/capacity
99
/sys/class/power_supply/BAT0/capacity_level
Normal
/sys/class/power_supply/BAT0/cycle_count
0
/sys/class/power_supply/BAT0/device
cat: /sys/class/power_supply/BAT0/device: Is a directory
/sys/class/power_supply/BAT0/energy_full
20320000
/sys/class/power_supply/BAT0/energy_full_design
23200000
/sys/class/power_supply/BAT0/energy_now
20280000
/sys/class/power_supply/BAT0/manufacturer
SANYO
/sys/class/power_supply/BAT0/model_name
45N1773
/sys/class/power_supply/BAT0/power
cat: /sys/class/power_supply/BAT0/power: Is a directory
/sys/class/power_supply/BAT0/power_now
0
/sys/class/power_supply/BAT0/present
1
/sys/class/power_supply/BAT0/serial_number
14432
/sys/class/power_supply/BAT0/status
Unknown
/sys/class/power_supply/BAT0/subsystem
cat: /sys/class/power_supply/BAT0/subsystem: Is a directory
/sys/class/power_supply/BAT0/technology
Li-ion
/sys/class/power_supply/BAT0/type
Battery
/sys/class/power_supply/BAT0/uevent
POWER_SUPPLY_NAME=BAT0
POWER_SUPPLY_STATUS=Unknown
POWER_SUPPLY_PRESENT=1
POWER_SUPPLY_TECHNOLOGY=Li-ion
POWER_SUPPLY_CYCLE_COUNT=0
POWER_SUPPLY_VOLTAGE_MIN_DESIGN=11100000
POWER_SUPPLY_VOLTAGE_NOW=12352000
POWER_SUPPLY_POWER_NOW=0
POWER_SUPPLY_ENERGY_FULL_DESIGN=23200000
POWER_SUPPLY_ENERGY_FULL=20320000
POWER_SUPPLY_ENERGY_NOW=20280000
POWER_SUPPLY_CAPACITY=99
POWER_SUPPLY_CAPACITY_LEVEL=Normal
POWER_SUPPLY_MODEL_NAME=45N1773
POWER_SUPPLY_MANUFACTURER=SANYO
POWER_SUPPLY_SERIAL_NUMBER=14432
/sys/class/power_supply/BAT0/voltage_min_design
11100000
/sys/class/power_supply/BAT0/voltage_now
12352000

Looks like according to this, the status is 'Unknown'.  If I'm understanding correctly, the commit [here](https://github.com/jaor/xmobar/commit/a2ed4f38c415259040d3e7513e72c1fab018feb1) is where you started reading from here instead of going off of the time left to charge and whether or not there was AC power? That's probably where the issue began.

**Comment by jaor** [comment]
> /sys/class/power_supply/BAT0/status
> Unknown

so the problem is this status.  for some reason, the status of the
battery isn't known (we didn't use this status directly in previous
versions, but that led to other bugs)... not sure if this is a hardware
or software issue: any ideas?

**Comment by dsemi** [comment]
Looks like /sys/class/power_supply/BAT1/status shows 'Charging' and 'Discharging' correctly. Not sure what the differences are between BAT0 and BAT1.

**Comment by dsemi** [comment]
According to [this](http://askubuntu.com/a/309146) (not sure how reliable it is), for kernel 3.19+ we should use BAT1.

**Comment by jaor** [comment]
and for BAT0 it's always Unknown?  does the `<watts>` field display
correctly for it (i.e., positive while charging, negative discharging
and zero while idle)?  if so, we could fallback to that when the battery
status is unknown...

**Comment by dsemi** [comment]
Yes, BAT0 is always Unknown. On BAT0 it looks like power_now is always 0.

**Comment by jaor** [comment]
so, if that's true, using `Battery ["BAT1"]` would solve the issue for you?

**Comment by dsemi** [comment]
Yeah, looks like it would. Maybe if BAT0 is Unknown you can fallback to BAT1? Assuming it exists.

**Comment by jaor** [comment]
that's more or less what the patch i've applied is doing...

**Comment by dsemi** [comment]
Looks like the [commit](https://github.com/jaor/xmobar/commit/37202c6552e1d3618d00e72ba033db8f51d6c4d1) fixed it.

**Comment by jaor** [comment]
thanks for checking, @dsemi. @YoEight is the patch also working for you?

**Comment by YoEight** [comment]
Hi, 

I checkout `master` code and tried but unfortunately it didn't work for me as I still get `N/A`.

My laptop has 2 batteries, it's a Thinkpad T450s.

Content of `/sys/class/power_supply/BAT0` is:

```
➜  xmobar git:(master) ✗ ls /sys/class/power_supply/BAT0
alarm           device              manufacturer  present        technology          voltage_now
capacity        energy_full         model_name    serial_number  type
capacity_level  energy_full_design  power         status         uevent
cycle_count     energy_now          power_now     subsystem      voltage_min_design
```

Output of `for f in /sys/class/power_supply/BAT0/*; do echo $f; cat $f; done` is:

```
➜  xmobar git:(master) ✗ for f in /sys/class/power_supply/BAT0/*; do echo $f; cat $f; done
/sys/class/power_supply/BAT0/alarm
946000
/sys/class/power_supply/BAT0/capacity
99
/sys/class/power_supply/BAT0/capacity_level
Normal
/sys/class/power_supply/BAT0/cycle_count
0
/sys/class/power_supply/BAT0/device
cat: /sys/class/power_supply/BAT0/device: Is a directory
/sys/class/power_supply/BAT0/energy_full
18930000
/sys/class/power_supply/BAT0/energy_full_design
23200000
/sys/class/power_supply/BAT0/energy_now
18840000
/sys/class/power_supply/BAT0/manufacturer
SANYO
/sys/class/power_supply/BAT0/model_name
45N1773
/sys/class/power_supply/BAT0/power
cat: /sys/class/power_supply/BAT0/power: Is a directory
/sys/class/power_supply/BAT0/power_now
0
/sys/class/power_supply/BAT0/present
1
/sys/class/power_supply/BAT0/serial_number
  689
/sys/class/power_supply/BAT0/status
Unknown
/sys/class/power_supply/BAT0/subsystem
cat: /sys/class/power_supply/BAT0/subsystem: Is a directory
/sys/class/power_supply/BAT0/technology
Li-ion
/sys/class/power_supply/BAT0/type
Battery
/sys/class/power_supply/BAT0/uevent
POWER_SUPPLY_NAME=BAT0
POWER_SUPPLY_STATUS=Unknown
POWER_SUPPLY_PRESENT=1
POWER_SUPPLY_TECHNOLOGY=Li-ion
POWER_SUPPLY_CYCLE_COUNT=0
POWER_SUPPLY_VOLTAGE_MIN_DESIGN=11100000
POWER_SUPPLY_VOLTAGE_NOW=12310000
POWER_SUPPLY_POWER_NOW=0
POWER_SUPPLY_ENERGY_FULL_DESIGN=23200000
POWER_SUPPLY_ENERGY_FULL=18930000
POWER_SUPPLY_ENERGY_NOW=18840000
POWER_SUPPLY_CAPACITY=99
POWER_SUPPLY_CAPACITY_LEVEL=Normal
POWER_SUPPLY_MODEL_NAME=45N1773
POWER_SUPPLY_MANUFACTURER=SANYO
POWER_SUPPLY_SERIAL_NUMBER=  689
/sys/class/power_supply/BAT0/voltage_min_design
11100000
/sys/class/power_supply/BAT0/voltage_now
12310000
```

Thanks for your time.

**Comment by YoEight** [comment]
Don't know if it can help but here's what `acpi` gives me:

```
➜  xmobar git:(master) ✗ acpi
Battery 0: Discharging, 9%, 00:16:09 remaining
Battery 1: Unknown, 99%
```

**Comment by jaor** [comment]
@YoEight, that output is rather puzzling, because your /status for BAT0 is "Unknown", yet acpi is reporting unknown for battery 1... what's the result of the for loop for BAT1? does unsing `BatteryN ["BAT1"]` work better?

**Comment by YoEight** [comment]
`BatteryN ["BAT1"]` doesn't work however `BatteryP ["BAT1"]` does.

**Comment by YoEight** [comment]
and by doesn't work I mean `xmobar` doesn't even display something, the whole bar is blank.

**Comment by YoEight** [comment]
Maybe something worth to be mentioned. My linux boot in legacy mode. I remember some month ago I didn't have any issue when using `EFI` mode. Really don't know if it's related.

**Comment by jaor** [comment]
maybe that's what causing the Unknown in the /status file... is it also
Unknown for BAT1?

**Comment by dsemi** [comment]
Ok mine isn't completely working with the fix. If I let the battery discharge, the status for BAT1 always shows Discharging as it should. If I then plug in the cable, it changes to Charging and then Full as it should. However, if while the battery is full, I unplug and then reinsert the power cord, the status becomes Unknown and appears as N/A.

**Comment by dsemi** [comment]
FWIW I also have a lenovo. T440s and it does have 2 batteries.

**Comment by YoEight** [comment]
By any chance @dsemi, you boot in legacy or EFI mode ?

**Comment by dsemi** [comment]
EFI

**Comment by jaor** [comment]
so it looks like one cannot trust /status, no? it seems to go to Unkown
whe it feels like it :) when that status is Unknown, does <watts> show a
good value? (if so, i could fallback to the sign of that to infer the
status of the battery).

**Comment by YoEight** [comment]
When using `Battery`, I got the value `-4` with `<watts>`. With `BatteryP ["BAT1"]`, I got  `-0`

**Comment by YoEight** [comment]
As @dsemi figured it out, `xmobar-0.23.1` Battery plugin works in my setup.

**Comment by jaor** [comment]
Battery is the one also giving you correct <watts>, right?

**Comment by YoEight** [comment]
Can't tell as it gives me a negative number. Does it means it consumes 4 watts ? If yes, `Battery` is the one giving me the right <watts>

**Comment by jaor** [comment]
it should mean that the battery is discharging at that rate, so if you were consuming battery that sounds right, yes.  if you plug the computer to AC, the number should turn positive (and usually be higher)...

**Comment by YoEight** [comment]
When plugged to AC, it's around 21 so `Battery` is the good one.

**Comment by jaor** [comment]
okay, so it seems that's a good indicator of the battery's status.  i've
just pushed a patch that uses it whenever the status of the batteries is
Unknown.  could you please give it a try?

**Comment by dsemi** [comment]
It shows charging and discharging properly, but it never says when it's full.

**Comment by jaor** [comment]
oops, i was reporting Idle instead of Full in that case.  i think i've
fixed it this time, sorry!

**Comment by YoEight** [comment]
Tried the last `master` version. When I no longer have `N/A`, now I unable to know how many time left my battery has. It's always  displaying `Charged`, even if AC is unplugged

**Comment by jaor** [comment]
and what do you see if you use <timeleft> directly in the main template
instead of only for the case of the battery charging?

**Comment by jaor** [comment]
i'm assuming both your BAT0/status and BAT1/status say Unknown, right?
i've pushed a change that uses the exact same criteria to decide the
battery status as it did in previous versions in that case.  so, if both
BAT0/status and BAT1/status report Unknown to you, this one should
really behave as it used to... would you mind giving it yet another try?
(sorry for being free with your time, but this is hard to debug on my
side because my statuses are just fine all the time).

**Comment by dsemi** [comment]
Looks to be working for me now.

**Comment by jaor** [comment]
since there was no further feedback from @YoEight, i'll assume this one is working also from him and close the issue.  @YoEight, please feel free to reopen it if you still have problems.

**Comment by dearieme** [comment]
I'm pretty sure I'm having the same issue, unfortunately I don't have a Haskell build environment to test master, will a release be tagged soon? Thanks

**Comment by jaor** [comment]
i've just released 0.24.3 with the latest fixes.  thanks!

**Comment by rig0rmortis** [comment]
I'm on a ThinkPad T460s (also 2 batteries), I'm running `xmobar` version `0.24.3-2` from the ArchLinux repositories. I've tried explicitly using `BatteryP` with `[BAT0, BAT1]` as an argument, and also just `Battery` but once my battery is fully charged I see my status as `Charged` and then it jumps down to `50%`, so it doesn't seem to be reading the second battery.

Here's output from acpi:
```
Battery0: Full, 100%
Battery1: Discharging, 31%, 01:08:20 remaining
```

Below is the battery section from my `.xmobarrc` (note that I have also tried `Run Battery` without specifying the batteries explicitly) 
```
, Run BatteryP    ["BAT0", "BAT1"]
                           [ "--template", "Batt: <acstatus>"
                           , "--Low"       , "10"    -- units: %
                           , "--High"      , "60"    -- units: %
                           , "--" -- battery specific options
                                   -- discharging status
                                   , "-o" , "<left>% (<timeleft>)"
                                   , "-O" , "Charging <left>%"
                                   , "-i" , "Charged"
                            ] 50
```

Have I just misconfigured my battery plugin or am I encountering a similar issue?

Thanks
